src/typemap.c

```diff
@@ -625,17 +615,14 @@ static jl_typemap_entry_t *jl_typemap_assoc_by_type_(jl_typemap_entry_t *ml, jl_
             else if (ml->issimplesig && !typesisva)
                 ismatch = sig_match_by_type_simple(jl_svec_data(types->parameters), n,
                                                    ml->sig, lensig, ml->va);
-            else if (ml->tvars == jl_emptysvec)
-                ismatch = jl_tuple_subtype(jl_svec_data(types->parameters), n, ml->sig, 0);
-            else if (penv == NULL) {
-                ismatch = jl_type_match((jl_value_t*)types, (jl_value_t*)ml->sig) != (jl_value_t*)jl_false;
-            }
+            else if (ml->tvars == jl_emptysvec || penv == NULL)
+                ismatch = jl_subtype((jl_value_t*)types, (jl_value_t*)ml->sig);
             else {
                 // TODO: this is missing the actual subtype test,
                 // which works currently because types is typically a leaf tt,
                 // or inexact is set (which then does a sort of subtype test via jl_types_equal)
                 // but this isn't entirely general
-                jl_value_t *ti = jl_lookup_match((jl_value_t*)types, (jl_value_t*)ml->sig, penv, ml->tvars);
+                jl_value_t *ti = jl_lookup_match((jl_value_t*)types, (jl_value_t*)ml->sig, penv);
```

Comment:
currently, we have been instead testing that `intersect(types, ml->sig) == types` because that is implied by `types <: ml->sig`, and the returned "undecidable" if the result is inconclusive. If this is changed to `jl_subtype_env`, it should be correct to also remove the second `if (inexact)` test.

the other usages of `inexact` look like those callers are calling the wrong function here. I think it looks like they should have been using `jl_matching_methods` (intersection lookup) with a limit of 1 instead of calling `jl_typemap_assoc_by_type` (subtype lookup).